components/common-go/baseserver/options.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package baseserver
+
+import (
+	"fmt"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+type config struct {
+	logger *logrus.Entry
+
+	// hostname is the hostname on which our servers will listen.
+	hostname string
+	// grpcPort is the port we listen on for gRPC traffic
+	grpcPort int
+	// httpPort is the port we listen on for HTTP traffic
+	httpPort int
+
+	// closeTimeout is the amount we allow for the server to shut down cleanly
+	closeTimeout time.Duration
+}
+
+func defaultConfig() *config {
+	return &config{
+		logger:       log.New(),
+		hostname:     "localhost",
+		httpPort:     9000,
+		grpcPort:     9001,
+		closeTimeout: 5 * time.Second,
+	}
+}
+
+type Option func(cfg *config) error
+
+func WithHostname(hostname string) Option {
+	return func(cfg *config) error {
+		cfg.hostname = hostname
+		return nil
+	}
+}
+
+func WithHTTPPort(port int) Option {
+	return func(cfg *config) error {
+		if port <= 0 {
+			return fmt.Errorf("http port must be greater than 0, got: %d", port)
+		}
+
+		cfg.httpPort = port
+		return nil
+	}
+}
+
+func WithGRPCPort(port int) Option {
+	return func(cfg *config) error {
+		if port < 0 {
```

Comment:
nit: Should it be ?
```suggestion
		if port <= 0 {
```

Reply:
This is actually a bug in the HTTP Port option. When you specify port `0`, `net.Listen` actually allocates an open port. I'll fix this in a follow-up PR to avoid tangling you in the review again. Thanks for flagging!